### PR TITLE
Add examples with an ILU(0) preconditioner

### PR DESCRIPTION
**Pull request by amontoison**
`DIOM` and `DQGMRES` work well with ILU(0) preconditioners.
I opened the pull request https://github.com/mcovalt/ILU0.jl/pull/3 in the hope that ILU0 becomes a registered package.

Should be only merged after https://github.com/mcovalt/ILU0.jl/pull/3.

### CODE REVIEW

File: examples/test_diom.jl [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+@printf("nnz(ILU) / nnz(A): %7.1e\n", nnz(F) / nnz(A))
+
+# Solve Ax = b with DIOM and an ILU(0) preconditioner
+yM = zeros(N)
+yN = zeros(N)
+yP = zeros(N)
+opM = LinearOperator(Float64, N, N, false, false, y -> forward_substitution(yM, F, y))
+opN = LinearOperator(Float64, N, N, false, false, y -> backward_substitution(yN, F, y))
+opP = LinearOperator(Float64, N, N, false, false, y -> ldiv!(yP, F, y))
```
Comment (dpo) [inline examples/test_diom.jl:26]: Let's factor out the problem setup, which is common to the two tests, to avoid duplication.

**Review by dpo — CHANGES_REQUESTED** [review]
Thanks! Could you please show the output of the examples?

**Comment by amontoison** [comment]
```julia
nnz(ILU) / nnz(A): 1.1e+00
[Without preconditioning] Residual norm:  4.7e-06
[Without preconditioning] Number of iterations: 239
[Split preconditioning] Residual norm:  3.6e-06
[Split preconditioning] Number of iterations:  62
[Left preconditioning] Residual norm:  4.2e-06
[Left preconditioning] Number of iterations:  62
[Right preconditioning] Residual norm:  3.3e-06
[Right preconditioning] Number of iterations:  62
```

**Comment by dpo** [comment]
Thanks. With ILU0, don't we expect nnz(ILU) = nnz(A)?

**Comment by amontoison** [comment]
I did the same remark myself yesterday and I suppose that the diagonal of `L` and `U` is not only composed of 1 (unlike Doolittle or Crout LU factorization) and the memory used for storing the diagonal of `A` is duplicate.

**Comment by amontoison** [comment]
I was wrong, I did a mistake when I implemented `nnz` function.
https://github.com/mcovalt/ILU0.jl/pull/3/files#diff-3897f5bfe01078dd825e0c8199fc12d59f1248306a331ed9b5683c5163d5ed8dR220

File: examples/test_dqgmres.jl [hunk 2]
```diff
@@ -0,0 +1,51 @@
+using Krylov, LinearOperators, ILU0
```
Comment (mcovalt) [inline examples/test_dqgmres.jl:1]:
```suggestion
using Krylov, LinearOperators, ILUZero
```

Package rename. See https://github.com/mcovalt/ILUZero.jl/commit/c4de1f0b5d405d917a13d85c75073f939429000a

File: examples/test_dqgmres.jl [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+@printf("nnz(ILU) / nnz(A): %7.1e\n", nnz(F) / nnz(A))
+
+# Solve Ax = b with DQGMRES and an ILU(0) preconditioner
+# Remark: DIOM can be used in the same way
+yM = zeros(N)
+yN = zeros(N)
+yP = zeros(N)
+opM = LinearOperator(Float64, N, N, false, false, y -> forward_substitution(yM, F, y))
+opN = LinearOperator(Float64, N, N, false, false, y -> backward_substitution(yN, F, y))
```
Comment (mcovalt) [inline examples/test_dqgmres.jl:26]:
```suggestion
opM = LinearOperator(Float64, N, N, false, false, y -> forward_substitution!(yM, F, y))
opN = LinearOperator(Float64, N, N, false, false, y -> backward_substitution!(yN, F, y))
```

Function rename. See https://github.com/mcovalt/ILUZero.jl/commit/a32ddc54ea97328e55de5ea1e901301614ec10fb

**Review by mcovalt — COMMENTED** [review]
I've initiated package registration: https://github.com/JuliaRegistries/General/pull/23415#issuecomment-714010268

I made some modifications to the [ILUZero.jl](https://github.com/mcovalt/ILUZero.jl) package that this PR depends on. This review suggests changes I believe will work with these modifications. I've not tested this, however.

**Comment by amontoison** [comment]
Thanks @mcovalt !
I updated the examples and it works.
I tested with an other example, the discretization of a Poisson PDE in polar coordinates and the results are :
```
[Without preconditioning] Residual norm:  1.4e-06
[Without preconditioning] Number of iterations: 5000 (maximum number of iterations)
[Split preconditioning] Residual norm:  7.4e-09
[Split preconditioning] Number of iterations: 129
[Left preconditioning] Residual norm:  8.5e-08
[Left preconditioning] Number of iterations:  64
[Right preconditioning] Residual norm:  1.5e-08
[Right preconditioning] Number of iterations: 168
```

**Comment by dpo** [comment]
Is nnz(L+U) = nnz(A) now?

**Comment by amontoison** [comment]
Yes, I fixed it. https://github.com/mcovalt/ILUZero.jl/commit/01340a2f3909a0c3eb1eb81e5f0f76bf09fc4e8b
When ILUZero will be registered (in 3 days), I will add `ILUZero` in the `examples/Project.toml`.

**Comment by dpo** [comment]
Great! Excellent work, guys!

**Comment by dpo** [comment]
Thank you.

**Pull request merged**